# Naive Bayes case study 

CSV file which tells which of the users purchased/not purchased a particular product.
I have used a Naive Bayes model to predict the purchase based on certain variables like gender, Age and Estimated Salary 

## Importing libraries 

In [96]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 

## Importing dataset

https://www.kaggle.com/rakeshrau/social-network-ads

In [97]:
dataset = pd.read_csv('C:/Users/binayak/Videos/Machine Learning/Classification/Naive Bayes/Social_Network_Ads.csv')

In [98]:
dataset

User ID  Gender  Age  EstimatedSalary  Purchased
0    15624510    Male   19            19000          0
1    15810944    Male   35            20000          0
2    15668575  Female   26            43000          0
3    15603246  Female   27            57000          0
4    15804002    Male   19            76000          0
..        ...     ...  ...              ...        ...
395  15691863  Female   46            41000          1
396  15706071    Male   51            23000          1
397  15654296  Female   50            20000          1
398  15755018    Male   36            33000          0
399  15594041  Female   49            36000          1

[400 rows x 5 columns]

In [99]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User ID          400 non-null    int64 
 1   Gender           400 non-null    object
 2   Age              400 non-null    int64 
 3   EstimatedSalary  400 non-null    int64 
 4   Purchased        400 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 14.1+ KB


In [100]:
dataset.describe()

User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000

## Separate the dependent variable from independent 

In [101]:
X = dataset.iloc[:,1:-1].values
y = dataset.iloc[:,-1].values

In [102]:
X

array([['Male', 19, 19000],
       ['Male', 35, 20000],
       ['Female', 26, 43000],
       ...,
       ['Female', 50, 20000],
       ['Male', 36, 33000],
       ['Female', 49, 36000]], dtype=object)

In [103]:
y

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1,

## Categorical variables : Encoded

In [104]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

ct = ColumnTransformer(transformers=[('encoder',OneHotEncoder(),[0])],
                       remainder='passthrough')

In [105]:
X= np.array(ct.fit_transform(X))

In [106]:
X

array([[0.0, 1.0, 19, 19000],
       [0.0, 1.0, 35, 20000],
       [1.0, 0.0, 26, 43000],
       ...,
       [1.0, 0.0, 50, 20000],
       [0.0, 1.0, 36, 33000],
       [1.0, 0.0, 49, 36000]], dtype=object)

## Split into Training and Test set

In [107]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=1/4)

In [108]:
print('The shape of X_train: {}; y_train: {}; X_test:{}; y_test:{}'.
      format(X_train.shape,y_train.shape,X_test.shape,y_test.shape))

The shape of X_train: (300, 4); y_train: (300,); X_test:(100, 4); y_test:(100,)


## Feature Scaling 

In [109]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()
X_train[:,2:] = ss.fit_transform(X_train[:,2:])
X_test[:,2:] = ss.transform(X_test[:,2:])

In [110]:
X_train[:10,:]

array([[0.0, 1.0, -0.20055268252278, -0.47501879453257817],
       [1.0, 0.0, -0.1042873949118457, 2.009694899945523],
       [1.0, 0.0, 0.9546307688084315, 1.8343033450411865],
       [0.0, 1.0, 0.08824318031002287, 1.1035051996064509],
       [0.0, 1.0, -0.20055268252278, 0.8988817188847249],
       [1.0, 0.0, 1.5322224944740372, -1.2350488657847032],
       [1.0, 0.0, -0.008022107300911418, 0.34347512835432575],
       [0.0, 1.0, -0.29681797013371425, -1.2058169399673138],
       [1.0, 0.0, 1.9172836449177744, -0.8842657559760302],
       [1.0, 0.0, 0.7621001935865629, -1.351976569054261]], dtype=object)

In [111]:
X_test[:10,:]

array([[1.0, 0.0, -0.29681797013371425, -1.3227446432368715],
       [1.0, 0.0, -0.5856138329665171, 2.3897099355715854],
       [1.0, 0.0, -0.29681797013371425, -0.6211784236195254],
       [1.0, 0.0, -1.4520014214649257, -0.06577183308912621],
       [0.0, 1.0, 1.82101835730684, 0.1680835734499892],
       [0.0, 1.0, -0.008022107300911418, -0.21193146217607334],
       [0.0, 1.0, -0.29681797013371425, 0.19731549926737862],
       [0.0, 1.0, -0.3930832577446486, 1.3665925319629557],
       [0.0, 1.0, 0.8583654811974971, 1.0742732737890615],
       [0.0, 1.0, -1.1632055586321228, 0.34347512835432575]], dtype=object)

## Training Naive Bayes on training set
class sklearn.naive_bayes.GaussianNB(*, priors=None, var_smoothing=1e-09)


In [112]:
from sklearn.naive_bayes import GaussianNB
nbc = GaussianNB()
nbc.fit(X_train,y_train)

GaussianNB()

## Predict purchase on test set

In [113]:
y_pred = nbc.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),
                    axis=1))

[[0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [0 1]]


## Making a Confusion Matrix & calculate Accuracy Score

In [114]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)
print(cm)

from sklearn.metrics import accuracy_score
a_score = accuracy_score(y_test,y_pred)
print(a_score)

[[61  6]
 [ 5 28]]
0.89


61 : correct prediction of class 0 

28 : correct prediction of class 1 

Hence 89% are correct (since test set has 100 observations)

Thus,

6 : incorrect prediction of class 1

5 : incorrect prediction of class 0
